src/ingestion/utils.py

```diff
@@ -25,21 +33,174 @@ def convert_html_to_markdown(html_file: str) -> str:
     return markdown_content
 
 
-def chunk_text(text: str, chunk_size: int = 300) -> List[str]:
+def num_tokens_from_string(string: str, encoding_name: str = 'cl100k_base') -> int:
+    """
+    Function that returns the number of tokens in a text string.
+    
+    Args:
+    string (str): The text for which we want to calculate the number of embeddings.
+    encoding_name (str): The name of the encoding model. Default 'cl100k_base'.
+
+    Returns:
+        int: Number of tokens of the input string.
+    """
+    # https://stackoverflow.com/questions/76106366/how-to-use-tiktoken-in-offline-mode-computer
+    encoding = tiktoken.get_encoding(encoding_name)
+    num_tokens = len(encoding.encode(string))
+    return num_tokens
+
+
+def merge_chunks(docs: List[Document], threshold: int) -> List[Document]:
```

Comment:
Secondo me questa funzione è un po' troppo complicata. Se ho capito l'intenzione, semplificandola un po' si possono anche evitare comportamenti un po' anomali (vedi esempi in codice sotto).
Idee:
- semplificare il flusso: es. confronto solo forward, ma con un loop
- se il threshold come da docstring si riferisce al max combined token count, occhio: nella funzione attuale il cfr è fatto con il token count singolo.

Proposta di una implementazione alternativa (contestabile!):
- loop su tutti i doc della lista come già fatto
- al doc i-esimo, un inner loop su tutti i doc successivi: ad ogni step di tale inner loop, si considera se merge possa essere fatto: in caso positivo lo si fa e si incrementa di 1 l'indice del loop, altrimenti si esce dall'inner loop, appendendo ciò che si è messo assieme e si va avanti.
mi aspetto che sia più semplice della funzione attuale, ma magari ne parliamo insime

```python
from langchain_core.documents import Document

from ingestion.utils import merge_chunks

print("Example 1: doc perso")
docs = [
    Document(metadata={"ch": "chapter1"}, page_content="Document A."),
    Document(metadata={"ch": "chapter1"}, page_content="Document B."),
    Document(metadata={"ch": "chapter1"}, page_content="Document C."),
]

threshold = 10
res = merge_chunks(docs, threshold=threshold)
print(res)

# risultato
# [Document(metadata={'ch': 'chapter1'}, page_content='Document B.\nDocument C.')]


print("Example 2: doc non uniti, anche se threshold molto ampio")
docs = [
    Document(metadata={"ch": "chapter1"}, page_content="Document A."),
    Document(metadata={"ch": "chapter1"}, page_content="Document B."),
    Document(metadata={"ch": "chapter1"}, page_content="Document C."),
    Document(metadata={"ch": "chapter1"}, page_content="Document D."),
]

threshold = 100
res = merge_chunks(docs, threshold=threshold)
print(res)

# risultato
# [Document(metadata={'ch': 'chapter1'}, page_content='Document A.\nDocument B.'), Document(metadata={'ch': 'chapter1'}, page_content='Document C.\nDocument D.')]

```
Fammi sapere se ti ritrovi o no :)